File: lib/cpp/src/leetcode/listnode.cpp
#include "listnode.h"

#include <sstream>
#include <vector>

ListNode* listNodeFrom(std::vector<int> input) {
    ListNode res(0);
    ListNode *node = &res;

    for (int n : input) { node->next = new ListNode(n); node = node->next; }

    return res.next;
}

bool operator==(const ListNode& lhs, const ListNode& rhs) {
    return (lhs.val == rhs.val)
        && !((lhs.next && !rhs.next) || (!lhs.next && rhs.next))
        && (!lhs.next || (*(lhs.next) == *(rhs.next)));
}

bool operator!=(const ListNode& lhs, const ListNode& rhs) {
    return !(lhs == rhs);
}

std::ostream& operator<<(std::ostream& os, const ListNode& value) {
    const ListNode *r = &value;

    os << "[";
    while (r) { os << r->val << ","; r = r->next; }
    os << "\b]";

    return os;
}

void destroy(ListNode* ln) {
    if (ln) { destroy(ln->next); delete ln; }
}

File: lib/rs/src/leetcode/listnode.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[rustfmt::skip]
    #[must_use]
    #[inline]
    pub fn new(val: i32) -> Self {
        ListNode {
            next: None,
            val
        }
    }

    /// Constructs a `ListNode` from `input`.
    #[must_use]
    pub fn from(input: Vec<i32>) -> Option<Box<Self>> {
        let mut res = Some(Box::new(ListNode::new(0)));
        let mut cur = res.as_mut();

        for n in input {
            if let Some(node) = cur {
                node.next = Some(Box::new(ListNode::new(n)));
                cur = node.next.as_mut();
            }
        }

        res?.next
    }
}
